common/src/main/java/com/zegelin/prometheus/netty/Server.java

```diff
@@ -62,27 +70,44 @@ public static void start(final List<InetSocketAddress> listenAddresses, final Ha
                 .channel(NioServerSocketChannel.class)
                 .childHandler(new ChildInitializer(harvester, helpExposition));
 
-        final List<Channel> channels;
         {
             final ImmutableList.Builder<Channel> builder = ImmutableList.builder();
 
             for (final InetSocketAddress listenAddress : listenAddresses) {
-                final Channel channel = bootstrap.bind(listenAddress).sync().channel();
-
-                builder.add(channel);
+                builder.add(bootstrap.bind(listenAddress).sync().channel());
             }
 
-            channels = builder.build();
+            Server.channels = builder.build();
         }
 
         if (logger.isInfoEnabled()) {
-            logger.info("cassandra-exporter started. Listening on {}", Joiner.on(", ").join(
+            logger.info("Cassandra Exporter Agent Server has started. Listening on {}", Joiner.on(", ").join(
                     listenAddresses.stream()
                             .map(a -> String.format("http://%s:%d", a.getHostString(), a.getPort()))
                             .iterator()
             ));
         }
+    }
+
+    public static void stop() {
+
+        if (Server.channels != null) {
 
-        // TODO: maybe return a future that the caller can sync on, to wait for the channels to shutdown?
+            for (Channel ch : channels) {
+                try {
+                    ch.closeFuture().sync();
+                } catch (InterruptedException e) {
+                    logger.debug("Closing of Cassandra Exporter Agent channel resulted in interrupted exception.", e);
+                }
+            }
+
+            Server.channels = null;
+
+            Server.eventLoopGroup.shutdownGracefully();
+
+            if (logger.isInfoEnabled()) {
```

Comment:
Usually it's [best practice to use parameterised messages](https://www.slf4j.org/faq.html#logging_performance), but in the case of the call to `logger.info(...)` in `start(...)` it's wrapped in a `logger.isInfoEnabled()` conditional. This is to avoid doing the work of mapping the addresses to strings and performing the string join. Good practice, but not like it really saves anything as `start(...)` only gets called once.

In this case, because it's a simple logging statement, the conditional is unnecessary (and theoretically involves doing _more_ work -- again, not like it really matters in this case).

Reply:
Thanks for the explanation, I was wondering why you are doing this so I just wanted to be consistent. I think that doing any conditionals here is just overkill as the effect is in practice imho very negligible. I removed this one and left the other one there.